[ORKTest] Extract task-creating methods to new TaskFactory class

rsanchezsaez opened the pull request:
At *4,813 lines of code*, `ORKTest`'s `MainViewController` class was getting a bit unmanageable. This PR extracts the task-creating methods to the new `TaskFactory` singleton class. This new class now includes the following state:

```
@property (nonatomic, copy) ORKConsentDocument *currentConsentDocument;
@property (nonatomic, strong) ORKTaskResult *embeddedReviewTaskResult;
```

which are used in the creation of some tasks.

Once this is merged, other minor improvements can be made to how buttons are wired to tasks in the test app, to generalize the task creation process so less code needs to be written when adding new tasks.

syoung-smallwisdom commented on Testing/ORKTest/ORKTest/TaskFactory.m:
```diff
@@ -0,0 +1,3622 @@
+//  ORKTest
+//
+//  Created by Ricardo Sanchez-Saez on 1/9/17.
+//  Copyright © 2017 ResearchKit. All rights reserved.
```
You forgot the copyright.

rsanchezsaez replied:
Done!

syoung-smallwisdom commented on Testing/ORKTest/ORKTest/TaskFactory.h:
```diff
@@ -0,0 +1,93 @@
+//  ORKTest
+//
+//  Created by Ricardo Sanchez-Saez on 1/9/17.
+//  Copyright © 2017 ResearchKit. All rights reserved.
```
copyright.

rsanchezsaez replied:
Thanks for spotting, fixed!

I took the chance to remove the *acute accent* (which I no longer use) from my name in all of the rest of files, I hope you don't mind.

syoung-smallwisdom commented on Testing/ORKTest/ORKTest/TaskFactory.m:
```diff
@@ -0,0 +1,3622 @@
+#pragma mark - Mapping identifiers to tasks
+
+- (id<ORKTask>)makeTaskWithIdentifier:(NSString *)identifier {
+    if ([identifier isEqualToString:DatePickingTaskIdentifier]) {
```
As a part of doing this, could you also alphabetically order the tasks? That would reduce the possibility of merge conflicts (instead of everyone adding to the bottom). Thanks!

rsanchezsaez replied:
See my comment below.

syoung-smallwisdom commented on Testing/ORKTest/ORKTest/TaskFactory.h:
```diff
@@ -0,0 +1,93 @@
+
+#define DefineStringKey(x) static NSString *const x = @#x
+
+DefineStringKey(ConsentTaskIdentifier);
```
Could you alphabetically order these to help with merge conflict resolution (and maintenance). Thanks!

rsanchezsaez replied:
That's a great idea. Do you think it's better to do it in a follow-up PR (to make this one easier to review)?

syoung-smallwisdom replied:
I guess I kind of figured that now, while you were doing the initial move to a new file, and while there are only three outstanding PR's that touch `ORKTest` would be the time to do it.

rsanchezsaezsc replied:
Sorry, do you mean you think is better to do it as part of this one or of a new one?

Also, if we want to hold on merging this PR before the other PRs get merged, that's no problem either.

rsanchezsaez replied:
Sorry, do you mean you think is better to do it as part of this one or of a new one?

Also, if we want to hold on merging this PR before the other PRs get merged, that's no problem either.

syoung-smallwisdom replied:
Sorry for the confusion. I meant that there aren't that many PRs right now that are affected by this so it is a good time to do it and yes, I think it's better to do the alphabetizing when you are moving the code b/c you've already created the new file and any new PRs will need to be changed to use the factory anyway.

rsanchezsaez commented:
@syoung-smallwisdom: I have added a few more sections to the `ORKTest` items, reordered the tests, and split the corresponding tasks into its own `TaskFactory` category files. They code's alphabetically sorted within each file.

I've also replaced some boilerplattery action and task switching code by selector-name calculation logic. I wouldn't do this in a production app for safety, but I think it's ok for a test app. I think it'll avoid needless code and it'll probably help people inserting the code for new tests in the appropriate places.

Let me know what you think, or if you have any other organizational suggestions. I'm happy to shuffle the categories around.

In the following days, I'll try to get rid of the *category identifier* constant, and I'll try to standardize the step identifier.

umerkhan-apple commented:
@rsanchezsaez Looks like there are merge conflicts and it seems that you are still updating some things in this PR. Can you ping me on this thread whenever you think the PR is ready to be merged?

syoung-smallwisdom commented:
@rsanchezsaez So good. Thank you. 👍

rsanchezsaez commented:
@umerkhan-apple Yeah, I'll make a few additional improvements. I'll let you done when I'm done. :)

rsanchezsaez commented:
@umerkhan-apple @syoung-smallwisdom This is now ready for review! :)

---

By the way, have a look at the new convenience block properties I added through a category to the steps, to be able to specify the behavior within within the `TaskFactory`.

```
typedef void (^StepViewControllerWillAppearBlockType)(ORKTaskViewController *taskViewController,
                                                      ORKStepViewController *stepViewController);
@property (nonatomic, copy) StepViewControllerWillAppearBlockType stepViewControllerWillAppearBlock;

typedef void (^StepViewControllerWillDisappearBlockType)(ORKTaskViewController *taskViewController,
                                                         ORKStepViewController *stepViewController,
                                                         ORKStepViewControllerNavigationDirection navigationDirection);
@property (nonatomic, copy) StepViewControllerWillDisappearBlockType stepViewControllerWillDisappearBlock;

typedef BOOL (^ShouldPresentStepBlockType)(ORKTaskViewController *taskViewController,
                                           ORKStep *step);
@property (nonatomic, copy) ShouldPresentStepBlockType shouldPresentStepBlock;
```

Do you think this is worth adopting as part of the main *ResearchKit* API? I felt this was quite useful, as it keeps the view controller step specific code close to the original step definition. On the downside, it's bundling UI logic to the `ORKStep`, which goes agains the original philosophy of an `ORKStep` being purely UI-agnostic data.

rsanchezsaez commented:
I've resolved the merge conflicts stemming from v1.5. This task is finished: please review if you can.  ;-)

I think this change is quite worthwhile, see https://github.com/ResearchKit/ResearchKit/pull/953/commits/724dd34bac8a69a8e44311923e17888ba5cfe863 for how little code was needed to re-add the *Stroop Active Task*.

rsanchezsaez commented:
Friendly ping @Pariecemckinney-apple @kelseydedoshka-apple. :-)

kelseydedoshka-apple commented:
@rsanchezsaez - this looks like a great idea. Pinging @akshay-yadav-apple as well as he has interfaced with the Stroop change recently and would love to get his feedback.

@Pariecemckinney-apple - please take a look at this for merging.

rsanchezsaez commented:
Thanks! I reckon the change is quit big for manual review. If you try reviewing the individual commits in chronological order, that may be easier to process. Also, note that this should only affect `ORKTest` sample code, and not any `ResearchKit` framework code (I reverted a silly copyright update commit, which I may submit later separately, to reduce review changes).
